# SingleCell

## Courses/Tutorials

1. Broad Institute: [Analysis of Single Cell RNAseq Data](https://broadinstitute.github.io/2019_scWorkshop/)
2. Hemberg Lab: [Analysis of Single Cell RNAseq Data](https://scrnaseq-course.cog.sanger.ac.uk/website/index.html)
3. ISMB: [SingleCellTranscriptomeTutorial](https://github.com/SingleCellTranscriptomics)
4. ontoProc: [Ontology interfaces for Bioconductor, with focus on cell type identification](http://www.bioconductor.org/packages/devel/bioc/vignettes/ontoProc/inst/doc/ontoProc.html)


## Tools

#### 

1. Seurat
2. Scanpy


#### Trajectory inference

1. dynverse: https://benchmark.dynverse.org/

#### Cell Type Annotation
1. SingleR (R pacakge)
2. cellassign
3. celaref
4. Moana (Python package, for PBMC only)
5. [CellO](https://github.com/deweylab/CellO): Cell Ontology-based classification (Python package)
6. [scMatch](https://github.com/forrest-lab/scMatch) (Python package)


## Databases

1. [Human Cell Atlas Data Portal](https://data.humancellatlas.org/)
2. [Single Cell Portal](https://portals.broadinstitute.org/single_cell)
3. Human Tumor Atlas Network

4. [Cell Markers](https://www.biolegend.com/en-us/cell-markers)
5. [CellMarker](http://biocc.hrbmu.edu.cn/CellMarker/#)


## Literatures

#### Trajectory Inference

1. [A comparison of single-cell trajectory inference methods. 2019. Nature Biotechnology](https://www.nature.com/articles/s41587-019-0071-9)

#### Cell Type Annotation

1. [Hierarchical cell type classification using mass, heterogeneous RNA-seq data from human primary cells](https://doi.org/10.1101/634097)


#### Reviews/Method Evaluation

1. [Bias, robustness and scalability in single-cell differential expression analysis](https://www.nature.com/articles/nmeth.4612)  

2. [Comparative analysis of differential gene expression analysis tools for single-cell RNA sequencing data](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-019-2599-6)  

3. [Single-Cell RNA-Seq Technologies and Related Computational Data Analysis](https://www.frontiersin.org/articles/10.3389/fgene.2019.00317/full)
